### PR TITLE
8256254: Convert vmIntrinsics::ID to enum class

### PR DESCRIPTION
**Pull request by iklam**
This PR is follows the same style as https://github.com/openjdk/jdk/pull/276, except this time I am converting `vmIntrinsics::ID` to `vmIntrinsicID`:

* Convert `vmIntrinsics::SID` to `enum class` to provide better type safety.
* Also, put this enum class in the top level, so it can be forward-declared. I.e., `enum class vmIntrinsicID : int;`. This avoids excessive inclusion of vmIntrinsics.hpp and vmSymbols.hpp (which were included indirectly by almost every hotspot source files).
  * vmIntrinsics.hpp: was included 805 times, now included 414 times
  * vmSymbols.hpp:  was included 805 times, now include 394 times
  * Note: more #include reduction will be done in [JDK-8256424](https://bugs.openjdk.java.net/browse/JDK-8256424)

Many files are changed, but most of them are minor
* Added missing dependencies of vmSymbols.hpp and/or vmIntrinsics.hpp
* safe conversion between vmIntrinsicID and integer types (see comments around `vmIntrinsics::as_int()`)

Since we have a lot of references like `vmIntrinsics::_invokeGeneric`, I added aliases like

```
static const vmIntrinsicID vmIntrinsics::_invokeGeneric = vmIntrinsicID::_invokeGeneric
```

so we don't need to change over a thousand `vmIntrinsics::XXX` to `vmIntrinsicID::XXX`.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8256254](https://bugs.openjdk.java.net/browse/JDK-8256254): Convert vmIntrinsics::ID to enum class

### Reviewers
 * [Claes Redestad](https://openjdk.java.net/census#redestad) (@cl4es - **Reviewer**) ⚠️ Review applies to 4f8f26923613df651916865128fd1aefd3ae4d5a
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 4f8f26923613df651916865128fd1aefd3ae4d5a

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1237/head:pull/1237`
`$ git checkout pull/1237`

### CODE REVIEW

File: src/hotspot/share/classfile/vmIntrinsics.hpp [hunk 1]
```diff
@@ -1057,10 +1072,33 @@ class vmIntrinsics : AllStatic {
...
+  // Convert an arbitrary vmIntrinsicID to int (checks validity):
+  //    vmIntrinsicID x = ...; int n = vmIntrinsics::as_int(x);
+  // Convert a known vmIntrinsicID to int (no need for validity check):
+  //    int n = static_cast<int>(vmIntrinsicID::_invokeGeneric);
+  static constexpr int as_int(vmIntrinsicID id) {
+    assert(is_valid_id(id), "must be");
+    return static_cast<int>(id);
+  }
+
+  static constexpr size_t number_of_intrinsics() {
```
Comment (cl4es) [inline src/hotspot/share/classfile/vmIntrinsics.hpp:1085]: Could this be returning int?

File: src/hotspot/share/jvmci/jvmciCompilerToVMInit.cpp [hunk 2]
```diff
@@ -161,7 +161,8 @@ void CompilerToVM::Data::initialize(JVMCI_TRAPS) {
 }
 
 JVMCIObjectArray CompilerToVM::initialize_intrinsics(JVMCI_TRAPS) {
-  JVMCIObjectArray vmIntrinsics = JVMCIENV->new_VMIntrinsicMethod_array(vmIntrinsics::ID_LIMIT - 1, JVMCI_CHECK_NULL);
+  int len = static_cast<int>(vmIntrinsics::ID_LIMIT) - 1;
```
Comment (cl4es) [inline src/hotspot/share/jvmci/jvmciCompilerToVMInit.cpp:164]: vmIntrinsics::number_of_intrinsics?

**Review by cl4es — APPROVED** [review]
Looks ok.

File: src/hotspot/share/jvmci/jvmciCompilerToVMInit.cpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/jvmci/jvmciCompilerToVMInit.cpp]: Thanks for the review. I have pushed a new version that addresses your comments.

**Review by cl4es — APPROVED** [review]

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by iklam** [comment]
Thanks @cl4es and @vnkozlov for the review. Passed tier1,tier2,builds-tier3,builds-tier4,builds-tier5
/integrate

**Pull request closed without merging**